Sources/Moho/ScopeCheck.swift

```diff
@@ -0,0 +1,576 @@
+          fatalError()
+        }
+        names.append(bindName)
+      }
+      return (names, tyExpr)
+    case let syntax as ImplicitTypedParameterSyntax:
+      let tyExpr = self.scopeCheckExpr(
+                      self.rebindArrows(syntax.ascription.typeExpr))
+      var names = [Name]()
+      for j in 0..<syntax.ascription.boundNames.count {
+        let name = Name(name: syntax.ascription.boundNames[j])
+        guard !self.isBoundVariable(name) else {
+          // If this declaration does not have a unique name, diagnose it and
+          // recover by ignoring it.
+          self.engine.diagnose(.nameShadows(name), node: syntax.ascription)
+          continue
+        }
+
+        guard let bindName = self.bindVariable(named: name) else {
+          fatalError("Name should be unique by now!")
+        }
+        names.append(bindName)
+      }
+      return (names, tyExpr)
+    default:
+      fatalError()
+    }
+  }
+
+  private func scopeCheckDeclaredPattern(_ syntax: DeclaredPattern) -> Pattern {
+    switch syntax {
+    case .wild:
+      return .wild
+    case let .variable(name):
+      guard let (localName, _) = self.lookupLocalName(name) else {
+        guard let name = self.bindVariable(named: name) else {
+          fatalError("Lookup failed, Name should be unbound!")
+        }
+        return .variable(QualifiedName(name: name))
+      }
+      return .variable(localName)
+    case let .constructor(name, ps):
+      return .constructor(name, ps.map(self.scopeCheckDeclaredPattern))
+    }
+  }
+
+  private func scopeCheckPattern(_ syntax: BasicExprListSyntax) -> [Pattern] {
+    assert(!syntax.isEmpty)
+
+    let pats = syntax.map(self.exprToDeclPattern)
+    var patterns = [Pattern]()
+    patterns.reserveCapacity(pats.count)
+    for p in pats.dropFirst() {
+      switch p {
+      case .wild:
+        patterns.append(.wild)
+      case .variable(let name):
+        guard let (localName, _) = self.lookupLocalName(name) else {
+          guard let name = self.bindVariable(named: name) else {
+            fatalError("Lookup failed, Name should be unbound!")
+          }
+          patterns.append(.variable(QualifiedName(name: name)))
+          continue
+        }
+        patterns.append(.variable(localName))
+      case let .constructor(n, ps):
+        let pats = ps.map(self.scopeCheckDeclaredPattern)
+        patterns.append(.constructor(n, pats))
+      }
+    }
+    return patterns
+  }
+
+  // FIXME: Remove this and really implement mixfix operators, dummy.
+  private func rebindArrows(_ syntax: ExprSyntax) -> ExprSyntax {
+    switch syntax {
+    case let syntax as ParenthesizedExprSyntax:
+      return syntax.withExpr(self.rebindArrows(syntax.expr))
+    case let syntax as ApplicationExprSyntax:
+      guard syntax.exprs.count > 1 else {
+        return syntax.exprs[0]
+      }
+
+      var precExprs = [BasicExprSyntax]()
+      for exprIdx in 0..<syntax.exprs.count {
```

Comment:
Comments for this would be really nice

Reply:
I intend to kill this.  It's a pile of hacks that's not worth the effort of documenting.

Reply:
Sounds good then